cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatDataset.java

```diff
@@ -0,0 +1,136 @@
+    NamespaceId.SYSTEM.dataset(Constants.ProgramHeartbeat.TABLE);
+  private static final byte[] ROW_KEY_SEPARATOR = Bytes.toBytes(":");
+
+  private final Table table;
+
+  public ProgramHeartbeatDataset(Table table) {
+    super("ignored", table);
+    this.table = table;
+  }
+
+  /**
+   * Static method for creating an instance of {@link ProgramHeartbeatDataset}.
+   */
+  public static ProgramHeartbeatDataset getOrCreate(DatasetContext datasetContext, DatasetFramework datasetFramework) {
+    try {
+      Table table = DatasetsUtil.getOrCreateDataset(datasetContext, datasetFramework, PROGRAM_HEARTBEAT_INSTANCE_ID,
+                                                    Table.class.getName(), DatasetProperties.EMPTY);
+      return new ProgramHeartbeatDataset(table);
+    } catch (DatasetManagementException | IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Write {@link RunRecordMeta} to heart beat dataset as value,
+   * with rowKey being <timestamp-in-seconds>:<program-run-id>
+   * @param runRecordMeta row value to write
+   * @param timestampInSeconds used for creating rowKey
+   */
+  public void writeRunRecordMeta(RunRecordMeta runRecordMeta, long timestampInSeconds) {
+    byte[] rowKey = createRowKey(timestampInSeconds, runRecordMeta.getProgramRunId());
+    table.put(rowKey, COLUMN_NAME, Bytes.toBytes(GSON.toJson(runRecordMeta)));
+  }
+
+  private byte[] createRowKey(long timestamp, ProgramRunId programRunId) {
+    return Bytes.add(Bytes.toBytes(timestamp), ROW_KEY_SEPARATOR, Bytes.toBytes(GSON.toJson(programRunId)));
+  }
+
+  /**
+   * Scan the table for the time range and return collection of latest {@link RunRecordMeta}
+   * we maintain the latest {@link RunRecordMeta} identified by {@link ProgramRunId},
+   * Since there can be more than one RunRecordMeta for the
+   * same runId due to multiple state changes and heart beat messages.
+   * @param startRowKey inclusive start rowKey
+   * @param endRowKey exclusive end rowKey
+   * @return collection of {@link RunRecordMeta}
+   */
+  public Collection<RunRecordMeta> scan(byte[] startRowKey, byte[] endRowKey) {
+    try (Scanner scanner = table.scan(startRowKey, endRowKey)) {
+      Row row;
+      Map<ProgramRunId, RunRecordMeta> runIdToRunRecordMap = new HashMap<>();
+      while ((row = scanner.next()) != null) {
+        RunRecordMeta runRecordMeta = GSON.fromJson(Bytes.toString(row.getColumns().get(COLUMN_NAME)),
+                                                    RunRecordMeta.class);
+        ProgramRunId runId = getProgramRunIdFromRowKey(row.getRow());
+        runIdToRunRecordMap.put(runId, runRecordMeta);
+      }
+
+      // since the serialized runRecordMeta doesn't have programRunId (transient), we will create and
+      // add the programRunId to RunRecordMeta and add to result list
+      List<RunRecordMeta> runRecordMetaList = new ArrayList();
+      runIdToRunRecordMap.forEach(
```

Comment:
you can also do this as `runIdToRunRecordMap.entrySet().stream().map().collect()`

Reply:
I updated to that pattern, however when using `forEach` pattern we provide a variable name for the key and value (runRecord, programRunId), which made the code more readable. is there any preference on which pattern to use ?